contracts/cw20-base/src/allowances.rs

```diff
@@ -0,0 +1,359 @@
+use cosmwasm_std::{
+    attr, Addr, Binary, BlockInfo, Deps, DepsMut, Env, MessageInfo, Order, Response, StdError,
+    StdResult, Storage, Uint128,
+};
+use cw20::{
+    AllAllowancesResponse, AllSpenderAllowancesResponse, AllowanceInfo, AllowanceResponse,
+    Cw20ReceiveMsg, Expiration, SpenderAllowanceInfo,
+};
+use cw20_allowances::Cw20Allowances;
+use cw_storage_plus::Bound;
+
+use crate::contract::Cw20Base;
+use crate::error::ContractError;
+use crate::state::{ALLOWANCES, ALLOWANCES_SPENDER, BALANCES, TOKEN_INFO};
+
+// settings for pagination
+const MAX_LIMIT: u32 = 30;
+const DEFAULT_LIMIT: u32 = 10;
+
+// this can be used to update a lower allowance - call bucket.update with proper keys
+pub fn deduct_allowance(
+    storage: &mut dyn Storage,
+    owner: &Addr,
+    spender: &Addr,
+    block: &BlockInfo,
+    amount: Uint128,
+) -> Result<AllowanceResponse, ContractError> {
+    let update_fn = |current: Option<AllowanceResponse>| -> _ {
+        match current {
+            Some(mut a) => {
+                if a.expires.is_expired(block) {
+                    Err(ContractError::Expired {})
+                } else {
+                    // deduct the allowance if enough
+                    a.allowance = a
+                        .allowance
+                        .checked_sub(amount)
+                        .map_err(StdError::overflow)?;
+                    Ok(a)
+                }
+            }
+            None => Err(ContractError::NoAllowance {}),
+        }
+    };
+    ALLOWANCES.update(storage, (owner, spender), update_fn)?;
+    ALLOWANCES_SPENDER.update(storage, (spender, owner), update_fn)
+}
+
+impl Cw20Allowances for Cw20Base {
+    type Error = ContractError;
+
+    /// Allows spender to access an additional amount tokens from the owner's (env.sender) account.
+    /// If expires is Some(), overwrites current allowance expiration with this one.
+    fn increase_allowance(
+        &self,
+        ctx: (DepsMut, Env, MessageInfo),
+        spender: String,
+        amount: Uint128,
+        expires: Option<Expiration>,
+    ) -> Result<Response, Self::Error> {
+        let (deps, env, info) = ctx;
+
+        let spender_addr = deps.api.addr_validate(&spender)?;
+        if spender_addr == info.sender {
+            return Err(ContractError::CannotSetOwnAccount {});
+        }
+
+        let update_fn = |allow: Option<AllowanceResponse>| -> Result<_, _> {
+            let mut val = allow.unwrap_or_default();
+            if let Some(exp) = expires {
+                if exp.is_expired(&env.block) {
+                    return Err(ContractError::InvalidExpiration {});
+                }
+                val.expires = exp;
+            }
+            val.allowance += amount;
+            Ok(val)
```

Comment:
Why do you force my brain to follow how is `val` change, instead of just:

```rust
match expires {
  let allow = unwrap_or_default();
  let allowance = allow.allowance + amount;
  Some(expires) if !exp.is_expired(&env.block) => Ok(AllowanceResponse {
    allowance,
    expires,
  }),
  None => Ok(AllownaceResponse {
    allowance,
    ..allow
  }),
  _ => Err(ContractError::InvalidExpiration {}),
}
```